README.md
driver-kata
===========
[![build](https://img.shields.io/travis/jbenner-radham/node-driver-kata.svg?style=flat-square)](https://travis-ci.org/jbenner-radham/node-driver-kata)
[![license](https://img.shields.io/github/license/jbenner-radham/node-driver-kata.svg?style=flat-square)](LICENSE)

> A code kata exercise.

Exercise Description
--------------------
Let's write some code to track driving history for people.

The code will process an input file. You can either choose to accept the input via stdin (e.g. if you're using Ruby
`cat input.txt | ruby yourcode.rb`), or as a file name given on the command line (e.g. `ruby yourcode.rb input.txt`).
You can use any programming language that you want. Please choose a language that allows you to best demonstrate your
programming ability.

Each line in the input file will start with a command. There are two possible commands.

The first command is Driver, which will register a new Driver in the app. Example:

`Driver Dan`

The second command is Trip, which will record a trip attributed to a driver. The line will be space delimited with the
following fields: the command (Trip), driver name, start time, stop time, miles driven. Times will be given in the
format of hours:minutes. We'll use a 24-hour clock and will assume that drivers never drive past midnight (the start
time will always be before the end time). Example:

`Trip Dan 07:15 07:45 17.3`

Discard any trips that average a speed of less than 5 mph or greater than 100 mph.

Generate a report containing each driver with total miles driven and average speed. Sort the output by most miles driven
to least. Round miles and miles per hour to the nearest integer.

Example input:

```
Driver Dan
Driver Alex
Driver Bob
Trip Dan 07:15 07:45 17.3
Trip Dan 06:12 06:32 21.8
Trip Alex 12:01 13:16 42.0
```

Expected output:

```
Alex: 42 miles @ 34 mph
Dan: 39 miles @ 47 mph
Bob: 0 miles
```

Install
-------
```sh
$ yarn # Or alternatively: `npm install`
```

Usage
-----
_**NOTE**: On Windows based systems you may need to invoke the CLI script via `node bin/cli.js`._

```sh
$ bin/cli.js --help

  A code kata exercise.

  Usage
      $ bin/cli.js <input file>

  Options
      --help, -h       Display this message.
      --version, -v    Display the application version.
```

For example output invoke the command `bin/cli.js example/input.txt`.


Testing
-------
```sh
$ yarn test # Or alternatively: `npm test`
```

Implementation Rationale
------------------------
Upon my initial review of the exercise description I felt that implementing classes for both of the command types was
the best starting point. This allows us to keep the logic and associated data organized.

I created the `Driver` and `Trip` classes that parse the input file lines passed into their constructors and assign the
data to the respective instance properties. Factory methods were added to each class to assist in a functional
implementation via `Array.prototype.map` as well as `Trip.prototype.isReported` for usage via chained
`Array.prototype.filter` invocations.

As each `Trip` belongs to a respective `Driver` instance I decided to add a `Driver.prototype.trips` property which is
populated via `Driver.prototype.addTrip` _(or the `Driver.prototype.addTrips` convenience method)_. Adding trips via
this method automatically updates the driver's miles driven and average speed. Lastly I added the
`Driver.prototype.toString` method to convert each driver instance into the textual representation that will be
aggregated into the final report. I felt using the `toString` nomenclature was the most semantic approach being that the
textual report generated is literally the data that the instance contains in the desired human readable format.

The final class implemented was `Report`. It effectively is a container for various `Driver` instances with the
respective `Report.prototype.toString` method to generate the driver report which is the primary objective of this
exercise.

I then added the primary `app` function in `lib/index.js` since it is the main library file. This function takes in
a single string input argument which is then split out into an array of lines which are put through a "functional
pipeline" to aggregate and filter the data into a `Report` instance at which point the value of
`Report.prototype.toString` is returned.

Lastly the executable `bin/cli.js` file imports the `app` function and utilizes the super handy
[meow](https://www.npmjs.com/package/meow) library to provide standard CLI functionality such as `--help` and
`--version` flags. At this point presuming that this script was invoked with a valid input filepath it will resolve and
read said file into a string which is then passed into the `app` function and prints the generated report to `stdout`.

See Also
--------
- [Yarn](https://yarnpkg.com/)

License
-------
The MIT License (Expat). See the [license file](LICENSE) for details.


CHANGELOG.md
Changelog
=========
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

0.1.0 - 2019-02-20
------------------
### Added
- Initial release.


test/Trip.spec.js
const { expect } = require('chai');
const Trip = require('../lib/Trip');

describe('Trip', function () {
    beforeEach(function () {
        this.trip = new Trip('Trip Dan 07:15 07:45 17.3');
    });

    it('is a function', function () {
        expect(Trip).to.be.a('function');
    });

    context('#constructor', function () {
        it('is a function', function () {
            expect(Trip.prototype.constructor).to.be.a('function');
        });

        it('sets the name property', function () {
            expect(this.trip.name).to.equal('Dan');
        });

        it('sets the start property', function () {
            const date = new Date(0);

            date.setUTCHours(7);
            date.setUTCMinutes(15);

            expect(this.trip.start.toISOString()).to.equal(date.toISOString());
        });

        it('sets the end property', function () {
            const date = new Date(0);

            date.setUTCHours(7);
            date.setUTCMinutes(45);

            expect(this.trip.end.toISOString()).to.equal(date.toISOString());
        });

        it('sets the duration property', function () {
            expect(this.trip.duration).to.equal(.5);
        });

        it('sets the miles property', function () {
            expect(this.trip.miles).to.equal(17.3);
        });

        it('throws a TypeError if not instantiated with a valid record', function () {
            const invalidInit = () => new Trip('Hello type error!');

            expect(invalidInit).to.throw(TypeError);
        });
    });

    context('.factory', function () {
        it('is a function', function () {
            expect(Trip.factory).to.be.a('function');
        });

        it('returns a new class instance', function () {
            expect(Trip.factory('Trip Dan 07:15 07:45 17.3')).to.be.an.instanceof(Trip);
        });

        it('caches the TypeError and returns null when passed an invalid record', function () {
            expect(Trip.factory('Invalid!')).to.equal(null);
        });
    });

    context('#isReported', function () {
        it('is a function', function () {
            expect(Trip.prototype.isReported).to.be.a('function');
        });

        it('returns a boolean', function () {
            expect(this.trip.isReported()).to.be.a('boolean');
        });

        it('returns false for a trip with an average speed of 4.9 mph', function () {
            const trip = new Trip('Trip Dan 07:15 08:15 4.9');

            expect(trip.isReported()).to.equal(false);
        });

        it('returns true for a trip with an average speed of 5 mph', function () {
            const trip = new Trip('Trip Dan 07:15 08:15 5.0');

            expect(trip.isReported()).to.equal(true);
        });

        it('returns true for a trip with an average speed of 60 mph', function () {
            const trip = new Trip('Trip Dan 07:15 08:15 60.0');

            expect(trip.isReported()).to.equal(true);
        });

        it('returns true for a trip with an average speed of 99.9 mph', function () {
            const trip = new Trip('Trip Dan 07:15 08:15 99.9');

            expect(trip.isReported()).to.equal(true);
        });

        it('returns false for a trip with an average speed of 100 mph', function () {
            const trip = new Trip('Trip Dan 07:15 08:15 100.0');

            expect(trip.isReported()).to.equal(false);
        });
    });

    context('.parseTime', function () {
        it('is a function', function () {
            expect(Trip.parseTime).to.be.a('function');
        });

        it('returns a date object', function () {
            expect(Trip.parseTime('07:15')).to.be.a('date');
        });

        it('returns a date seeded from the Unix epoch', function () {
            const date = Trip.parseTime('08:55');

            expect(date.toISOString().startsWith('1970-01-01')).to.be.true;
        });

        it('sets the UTC hours on the returned date', function () {
            const date = Trip.parseTime('08:55');
            const hours = date.getUTCHours();

            expect(hours).to.equal(8);
        });

        it('sets the UTC minutes on the returned date', function () {
            const date = Trip.parseTime('08:55');
            const minutes = date.getUTCMinutes();

            expect(minutes).to.equal(55);
        });
    });

    context('#getDuration', function () {
        it('is a function', function () {
            expect(Trip.prototype.getDuration).to.be.a('function');
        });

        it('returns a number', function () {
            const trip = new Trip('Trip Dan 07:15 07:45 17.3');
            const duration = trip.getDuration();

            expect(duration).to.be.a('number');
        });

        it('returns 1 for an hour long trip', () => {
            const trip = new Trip('Trip Dan 05:55 06:55 55.5');
            const duration = trip.getDuration();

            expect(duration).to.equal(1);
        });
    });
});


test/Driver.spec.js
const { expect } = require('chai');
const Driver = require('../lib/Driver');

describe('Driver', function () {
    beforeEach(function () {
        this.driver = new Driver('Driver Dan');
    });

    it('is a function', function () {
        expect(Driver).to.be.a('function');
    });

    context('#constructor', function () {
        it('is a function', function () {
            expect(Driver.prototype.constructor).to.be.a('function');
        });

        it('sets the name property', function () {
            expect(this.driver.name).to.equal('Dan');
        });

        it('sets the trips property to an empty array', function () {
            expect(this.driver.trips).to.eql([]);
        });

        it('throws a TypeError if not instantiated with a valid record', function () {
            const invalidInit = () => new Driver('Hello type error!');

            expect(invalidInit).to.throw(TypeError);
        });
    });

    context('.factory', function () {
        it('is a function', function () {
            expect(Driver.factory).to.be.a('function');
        });

        it('returns a new class instance', function () {
            expect(Driver.factory('Driver Dan')).to.be.an.instanceof(Driver);
        });

        it('catches the TypeError and returns null when passed an invalid record', function () {
            expect(Driver.factory('Invalid!')).to.equal(null);
        });
    });

    context('#getAverageSpeed', function () {
        it('is a function', function () {
            expect(Driver.prototype.getAverageSpeed).to.be.a('function');
        });

        it('returns the average miles per hour of the driver', function () {
            expect(this.driver.getAverageSpeed()).to.equal(0);
        });

        it('returns a number', function () {
            expect(this.driver.getAverageSpeed()).to.be.a('number');
        });
    });

    context('#addTrip', function () {
        beforeEach(function () {
            this.tripFixture = {
                duration: 1,
                miles: 60,
                name: this.driver.name
            };
        });

        it('is a function', function () {
            expect(Driver.prototype.addTrip).to.be.a('function');
        });

        it('adds a trip for the driver', function () {
            this.driver.addTrip(this.tripFixture);
            expect(this.driver.trips).to.eql([this.tripFixture]);
        });

        it('updates the average speed of the driver', function () {
            this.driver.addTrip(this.tripFixture);
            expect(this.driver.averageSpeed).to.equal(60);
        });

        it('updates the miles driven by the driver', function () {
            this.driver.addTrip(this.tripFixture);
            expect(this.driver.milesDriven).to.equal(this.tripFixture.miles);
        });

        it('throws a TypeError if not passed a trip object for the driver', function () {
            const tripFixture = { name: 'Totally Not Dan' };
            const invalidCall = () => this.driver.addTrip(tripFixture);

            expect(invalidCall).to.throw(TypeError);
        });
    });

    context('#addTrips', function () {
        beforeEach(function () {
            this.tripsFixture = [{
                duration: 1,
                miles: 60,
                name: this.driver.name
            }, {
                duration: .55,
                miles: 13.1,
                name: this.driver.name
            }];
        });

        it('is a function', function () {
            expect(Driver.prototype.addTrips).to.be.a('function');
        });

        it('adds trips for the driver', function () {
            this.driver.addTrips(this.tripsFixture);
            expect(this.driver.trips).to.eql(this.tripsFixture);
        });

        it('updates the average speed of the driver', function () {
            this.driver.addTrips(this.tripsFixture);
            expect(this.driver.averageSpeed).to.equal(this.driver.getAverageSpeed());
        });

        it('updates the miles driven by the driver', function () {
            this.driver.addTrips(this.tripsFixture);
            expect(this.driver.milesDriven).to.equal(73.1);
        });

        it('throws a TypeError if not passed trip objects for the driver', function () {
            const tripsFixture = [{ name: 'Totally Not Dan' }];
            const invalidCall = () => this.driver.addTrips(tripsFixture);

            expect(invalidCall).to.throw(TypeError);
        });
    });

    context('#toString', function () {
        it('is a function', function () {
            expect(Driver.prototype.toString).to.be.a('function');
        });

        it('returns a string', function () {
            expect(this.driver.toString()).to.be.a('string');
        });

        it('generates a report for a driver with no miles traveled', function () {
            expect(this.driver.toString()).to.equal('Dan: 0 miles');
        });
    });
});


test/Report.spec.js
const { expect } = require('chai');
const os = require('os');
const Driver = require('../lib/Driver');
const Report = require('../lib/Report');
const Trip = require('../lib/Trip');

describe('Report', function () {
    beforeEach(function () {
        const getDriversFixture = () => {
            const alex = new Driver('Driver Alex');
            const bob = new Driver('Driver Bob');
            const dan = new Driver('Driver Dan');

            alex.addTrips([
                new Trip('Trip Alex 12:01 13:16 42.0')
            ]);
            dan.addTrips([
                new Trip('Trip Dan 07:15 07:45 17.3'),
                new Trip('Trip Dan 06:12 06:32 21.8')
            ]);

            return [alex, bob, dan];
        };

        this.driversFixture = getDriversFixture();
        this.report = new Report();
    });

    it('is a function', function () {
        expect(Report).to.be.a('function');
    });

    context('#constructor', function () {
        it('is a function', function () {
            expect(Report.prototype.constructor).to.be.a('function');
        });

        it('sets the drivers property to an empty array', function () {
            expect(this.report.drivers).to.be.an('array').with.a.lengthOf(0);
        });
    });

    context('#addDriver', function () {
        beforeEach(function () {
            this.driverFixture = { name: 'Jane' };
        });

        it('is a function', function () {
            expect(Report.prototype.addDriver).to.be.a('function');
        });

        it('adds a driver the drivers array property', function () {
            this.report.addDriver(this.driverFixture);

            expect(this.report.drivers).to.eql([this.driverFixture]);
        });
    });

    context('#addDrivers', function () {
        beforeEach(function () {
            this.driversFixture = [
                { name: 'Jane' },
                { name: 'John' }
            ];
        });

        it('is a function', function () {
            expect(Report.prototype.addDrivers).to.be.a('function');
        });

        it('adds drivers the drivers array property', function () {
            this.report.addDrivers(this.driversFixture);

            expect(this.report.drivers).to.eql(this.driversFixture);
        });
    });

    context('#toArray', function () {
        it('is a function', function () {
            expect(Report.prototype.toArray).to.be.a('function');
        });

        it('returns an array', function () {
            expect(this.report.toArray()).to.be.a('array');
        });

        it('returns an array of the formatted reports', function () {
            const expected = [
                'Alex: 42 miles @ 34 mph',
                'Bob: 0 miles',
                'Dan: 39 miles @ 47 mph'
            ];

            this.report.addDrivers(this.driversFixture);
            expect(this.report.toArray()).to.eql(expected);
        });
    });

    context('#toString', function () {
        it('is a function', function () {
            expect(Report.prototype.toString).to.be.a('function');
        });

        it('returns a string', function () {
            expect(this.report.toString()).to.be.a('string');
        });

        it('returns a string of the formatted reports', function () {
            const expected = [
                'Alex: 42 miles @ 34 mph',
                'Bob: 0 miles',
                'Dan: 39 miles @ 47 mph'
            ].join(os.EOL);

            this.report.addDrivers(this.driversFixture);
            expect(this.report.toString()).to.equal(expected);
        });
    });
});


.eslintrc.js
/**
 * @see https://eslint.org/docs/rules/
 */
module.exports = {
    env: {
        es6: true,
        node: true
    },
    extends: 'eslint:recommended',
    parserOptions: {
        ecmaVersion: 9 // Same as the now deprecated 2018 option, see https://eslint.org/docs/user-guide/configuring#deprecated
    },
    root: true,
    rules: {
        'arrow-spacing': 'error',
        'block-spacing': 'error',
        camelcase: 'error',
        'comma-dangle': ['error', 'never'],
        'comma-spacing': ['error', { after: true, before: false }],
        'computed-property-spacing': ['error', 'never'],
        'consistent-return': 'off',
        'eol-last': ['error', 'always'],
        eqeqeq: ['error', 'smart'],
        indent: ['error', 4],
        'key-spacing': [
            'error',
            { afterColon: true, beforeColon: false, mode: 'strict' }
        ],
        'keyword-spacing': 'error',
        'linebreak-style': ['error', 'unix'],
        'max-len': ['error', { code: 120, comments: 120 }],
        'no-console': process.env.NODE_ENV === 'production' ? 'error' : 'off',
        'no-debugger': process.env.NODE_ENV === 'production' ? 'error' : 'off',
        'no-tabs': ['error', { allowIndentationTabs: false }],
        'no-trailing-spaces': 'error',
        'no-var': 'error',
        'object-curly-spacing': ['error', 'always'],
        'object-property-newline': [
            'error',
            { allowAllPropertiesOnSameLine: true }
        ],
        'prefer-const': [
            'error',
            { destructuring: 'any', ignoreReadBeforeAssign: false }
        ],
        'quote-props': ['error', 'as-needed'],
        quotes: ['error', 'single', { allowTemplateLiterals: true }],
        semi: ['error', 'always'],
        'sort-keys': [ 'error', 'asc', { caseSensitive: false, natural: true } ],
        'space-before-blocks': [
            'error',
            { classes: 'always', functions: 'always', keywords: 'always' }
        ],
        'space-before-function-paren': [
            'error',
            { anonymous: 'always', asyncArrow: 'always', named: 'never' }
        ],
        'space-infix-ops': ['error', { int32Hint: false }],
        'space-unary-ops': [ 'error', { nonwords: false, words: true } ],
        'spaced-comment': [ 'error', 'always' ],
        'switch-colon-spacing': ['error', { after: true, before: false }]
    }
};


lib/index.js
const _ = require('lodash');
const Driver = require('./Driver');
const Report = require('./Report');
const Trip = require('./Trip');

/**
 * @param {string} input - Input file contents.
 */
module.exports = function app(input) {
    const newlineRegex = /\r?\n/;
    const isNotEmptyStr = str => str !== '';
    const isNotNull = data => data !== null;
    const isTripReported = (trip) => trip.isReported();
    const report = new Report();

    const lines = input.split(newlineRegex)
        .map(_.trim)
        .filter(isNotEmptyStr);

    const trips = lines.map(Trip.factory)
        .filter(isNotNull)
        .filter(isTripReported);

    const drivers = _(lines).map(Driver.factory)
        .filter(isNotNull)
        .uniqBy('name')
        .sortBy('name')
        .valueOf();

    drivers.forEach(driver => {
        const isDriversTrip = (trip) => driver.name === trip.name;

        driver.addTrips(trips.filter(isDriversTrip));
    });

    report.addDrivers(drivers);

    return report.toString();
};


lib/Driver.js
/**
 * @property {string} name
 * @property {number} milesDriven
 * @property {?number} averageSpeed
 * @property {Trip[]} trips
 */
module.exports = class Driver {
    /**
     * Creates an instance of Driver.
     *
     * @param {string} record - A trip record e.g., `Driver Dan`.
     * @throws {TypeError} Record argument must be in the required syntax.
     */
    constructor(record) {
        const regex = /^Driver (\w+)$/;

        if (!regex.test(record)) {
            throw new TypeError('The record provided does not match the required syntax');
        }

        const [, name] = regex.exec(record);

        this.name = name;
        this.milesDriven = 0;
        this.averageSpeed = null;
        this.trips = [];
    }

    /**
     * A functional utility method for this class.
     *
     * @static
     * @param {string} record - A trip record e.g., `Driver Dan`.
     * @returns {?Driver}
     */
    static factory(record) {
        try {
            return new Driver(record);
        } catch (_) {
            return null;
        }
    }

    /**
     * Calculates the average speed based upon the trips added to the instance.
     *
     * @returns {number} Average miles per hour as a float.
     */
    getAverageSpeed() {
        const addDuration = (accumulator, trip) => accumulator + trip.duration;
        const addMiles = (accumulator, trip) => accumulator + trip.miles;
        const hoursDriven = this.trips.reduce(addDuration, 0);
        const milesDriven = this.trips.reduce(addMiles, 0);
        const averageSpeed = milesDriven / hoursDriven;

        return Number.isFinite(averageSpeed) ? averageSpeed : 0;
    }

    /**
     * Add a trip for the driver.
     *
     * @param {Trip} trip
     * @throws {TypeError} Type#name must match the name property.
     */
    addTrip(trip) {
        if (trip.name !== this.name) {
            throw new TypeError('The trip provided is not for this driver');
        }

        this.trips = [...this.trips, trip];
        this.averageSpeed = this.getAverageSpeed();
        this.milesDriven += trip.miles;
    }

    /**
     * Add trips for the driver.
     *
     * @param {Trip[]} trips
     * @throws {TypeError} Every Type#name must match the name property.
     */
    addTrips(trips) {
        trips.forEach(this.addTrip.bind(this));
    }

    /**
     * Generates a driver report e.g., `Dan: 39 miles @ 47 mph`.
     *
     * @returns {string}
     */
    toString() {
        const { name, milesDriven, averageSpeed } = this;
        const miles = `${Math.round(milesDriven)} miles`;
        const speed = averageSpeed !== null
            ? `@ ${Math.round(averageSpeed)} mph`
            : '';

        return `${name}: ${miles} ${speed}`.trim();
    }
};


lib/Report.js
const _ = require('lodash');
const os = require('os');

/**
 * @property {Driver[]} [drivers=[]]
 */
module.exports = class Report {
    /**
     * Creates an instance of Report.
     */
    constructor() {
        this.drivers = [];
    }

    /**
     * Add a driver object to the report and de-duplicates drivers if needed.
     *
     * @param {Driver} driver
     */
    addDriver(driver) {
        this.drivers = _([...this.drivers, driver])
            .uniqBy('name')
            .sortBy('name')
            .valueOf();
    }

    /**
     * Add driver objects to the report and de-duplicates drivers if needed.
     *
     * @param {Driver[]} drivers
     */
    addDrivers(drivers) {
        drivers.forEach(this.addDriver.bind(this));
    }

    /**
     * Get the formatted driver reports as an array.
     *
     * @returns {string[]} A collection of reports e.g., `['Alex: 42 miles @ 34 mph']`
     */
    toArray() {
        return this.drivers.map(driver => driver.toString());
    }

    /**
     * Get the formatted driver reports as a string.
     *
     * @returns {string} A formatted text report of the added drivers.
     */
    toString() {
        return this.toArray().join(os.EOL);
    }
};


lib/Trip.js
/**
 * @property {string} name
 * @property {Date} start
 * @property {Date} end
 * @property {number} duration
 * @property {number} miles
 */
module.exports = class Trip {
    /**
     * Creates an instance of Trip.
     *
     * @param {string} - A trip record e.g., `Trip Dan 07:15 07:45 17.3`
     * @throws {TypeError} Record argument must be in the required syntax.
     */
    constructor(record) {
        const regex = /^Trip (\w+) (\d{1,2}:\d{2}) (\d{1,2}:\d{2}) (\d+(?:.\d)?)$/;

        if (!regex.test(record)) {
            throw new TypeError('The record provided does not match the required syntax');
        }

        const [, name, start, end, miles] = regex.exec(record);

        this.name = name;
        this.start = Trip.parseTime(start);
        this.end = Trip.parseTime(end);
        this.duration = this.getDuration();
        this.miles = Number.parseFloat(miles);
    }

    /**
     * A functional utility method for this class.
     *
     * @static
     * @param {string} record - A trip record e.g., `Trip Dan 07:15 07:45 17.3`
     * @returns {?Trip}
     */
    static factory(record) {
        try {
            return new Trip(record);
        } catch (_) {
            return null;
        }
    }

    /**
     * Checks that a trip does not average a speed of less than 5 mph or greater than 100 mph.
     *
     * @returns {boolean}
     */
    isReported() {
        const speed = (distance, time) => distance / time;
        const { duration, miles } = this;
        const averageSpeed = speed(miles, duration);

        return averageSpeed >= 5 && averageSpeed < 100;
    }

    /**
     * Convert a trip event time to a date object.
     *
     * @static
     * @param {string} time - A 24-hour time in HH:mm format.
     * @returns {Date}
     */
    static parseTime(time) {
        const date = new Date(0);
        const [hours, minutes] = time.split(':');

        date.setUTCHours(hours);
        date.setUTCMinutes(minutes);

        return date;
    }

    /**
     * Get the duration of the trip in hours.
     *
     * @returns {number}
     */
    getDuration() {
        const millisecondsInAMinute = 60000;
        const minutesInAnHour = 60;
        const durationInMilliseconds = this.end - this.start;
        const durationInMinutes = durationInMilliseconds / millisecondsInAMinute;

        return durationInMinutes / minutesInAnHour;
    }
};
